xpdan/data_reduction_core.py

```diff
@@ -0,0 +1,576 @@
+        (horizontal). default is 0.99. set to None for no
+        correction.
+    mask_setting : str, ndarray optional
+        string for mask option. Valid options are 'default', 'auto' and
+        'None'. If 'default', mask included in metadata will be
+        used. If 'auto', a new mask would be generated from current
+        image. If 'None', no mask would be applied. If a ndarray of bools
+        use as mask. If a path/filename to a valid fit2d or numpy mask file
+        (with extensions of `.msk` or `.npy`) load the file and use that.
+        Predefined option is 'default'.
+    mask_dict : dict, optional
+        dictionary stores options for automasking functionality.
+        default is defined by an_glbl.auto_mask_dict.
+        Please refer to documentation for more details
+    save_image : bool, optional
+        option to save dark subtracted images. images will be
+        saved to the same directory of chi files. default is True.
+    root_dir : str, optional
+        path of chi files that are going to be saved. default is
+        the same as your image file
+    config_dict : dict, optional
+        dictionary stores integration parameters of pyFAI azimuthal
+        integrator. default is the most recent parameters saved in
+        xpdUser/conifg_base
+    image_data_key: str, optional
+        The key for the image data, defaults to `pe1_image`
+    kwargs :
+        addtional keywords to overwrite integration behavior. Please
+        refer to pyFAI.azimuthalIntegrator.AzimuthalIntegrator for
+        more information
+
+    Note
+    ----
+    complete docstring of masking functionality could be find in
+    ``mask_img``
+
+    customized mask can be assign to by kwargs (It must be a ndarray)
+    >>> integrate_and_save(mask_setting=my_mask)
+
+    See also
+    --------
+    xpdan.tools.mask_img
+    pyFAI.azimuthalIntegrator.AzimuthalIntegrator
+    """
+    if mask_dict is None:
+        mask_dict = {}
+    header_list = _prepare_header_list(headers)
+    ai = AzimuthalIntegrator()
+
+    total_rv_list_Q = []
+    total_rv_list_2theta = []
+
+    if not isinstance(path_append_keys, (list, tuple)):
+        path_append_keys = [path_append_keys]
+
+    for header in header_list:
+        header_rv_list_Q, header_rv_list_2theta = [], []
+        start = False
+        event = False
+        stop = False
+        for name, doc in db.restream(header, fill=True):
+            if name == 'start':
+                start = True
+                # config_dict
+                if config_dict is None:
+                    config_dict = _load_config(header)  # default dict
+                    if config_dict is None:  # still None
+                        raise RuntimeError(
+                            "Can't find calibration parameter under "
+                            "xpdUser/config_base/ or header metadata\n"
+                            "data reduction can not be perfomed.")
```

Comment:
This is likely because a run_calibration() was not carried out before the data were collected during data acquisition.\\
If you have  a calibration file, please rerun integrate_and_save() giving it a path to the calib file, e.g., integrate_and_save(header, config_base='<path to the folder with the file>', calib_config_name='<calibration file name>').   
If you have many files that take the same calibration file, define config_base and calib_config_name in your config file.